Compatibility to BrianHenryIE/strauss

swarnat opened the pull request:
**fix**: compatibility to [BrianHenryIE/strauss](https://github.com/BrianHenryIE/strauss)

This is a very special modification for only some users, but which are currently impossible to use this helpfull library

[BrianHenryIE/strauss](https://github.com/BrianHenryIE/strauss) library will prefix dependency classnames with a custom string to prevent dependency collision.

For example classname will be changed from **Less_Visitor** to **ModA_Less_Visitor**. 
So multiple versions of same library can be used in ecosystems, where external modules can be installed from different developers.

The modified line generate function calls from class names, by removing the "namespace" part of class name.
When there is a custom prefix, then this line is not working.

The small modification check for custom prefix and append them to classname replacement.
Not a perfect solution. Maybe there is a better one.

Krinkle commented:
I suggest to, instead of detecting a (generally non-existent) prefix, to focus only on extracting the suffix after "Less_Tree_". I prefer to 1) avoid encoding an assumption between Visitor and Tree, and 2) not support or maintain extra code for something we can't easily test for and would never happen outside of the code modification scenario, as such code is likely to seem as "unused code" and be break or be removed in the future by accident.

OptionalAnythingLess_Tree_Foo_Bar > Foo_Bar > FooBar

Changing the transform from prefix to suffix means:
* the result is the same for the normal case
* the logic remains specific to the Tree class, no assumption to Visitor
* no extra code. Instead we replace the current code.same level of complexity instead of more.
* Greater flexibility, as it allows any custom Tree or Visitor to work with the same code, not all or nothing.

swarnat commented:
I would agree with you, because to do a simple _str_replace_ sounds a little bit unsecure in my case. 
But yeah. Without your opinion and explanation I don't see a way to implement it, because there are classes like **Less_Tree_Dimension** and **Less_Tree_Mixin_Definition**. 
So a simple split by underscore is not possible. 

But yeah, because the "Less_Tree" must removed, I can also split by that string and take the last element.
This do not fulfill the "no extra code", but the other solution I see would be a _substr_ combined by _str_pos_ + strlen("Less_Tree")
I do not see any benefits different to the split.
My commit: 253969b9bda430c2450c4f2c27671aefa92c3f83

I already moved the diff to gerrit, because the bot don't like that here. :)
https://gerrit.wikimedia.org/r/c/mediawiki/libs/less.php/+/1069631

Krinkle commented:
Thanks @swarnat. I left a suggestion there.

Krinkle commented:
@swarnat FYI, I've merged your patch at https://gerrit.wikimedia.org/r/c/mediawiki/libs/less.php/+/1069631. It will be part of the next release.
